keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,1699 @@
++       // containers to be assigned to.
++       ContainerQoSResources []QoSResourceInfo
++}
+
++// QoSResourceInfo contains information about one QoS resource type.
++type QoSResourceInfo struct {
++       // Name of the resource.
++       Name QoSResourceName
++       // Mutable is set to true if the resource supports in-place updates.
++       Mutable bool
++       // Classes available for assignment.
++       Classes []QoSResourceClassInfo
++}
+
++// QoSResourceClassInfo contains information about single class of one QoS
++// resource.
++type QoSResourceClassInfo struct {
++       // Name of the class.
++       Name string
++       // Capacity is the number of maximum allowed simultaneous assignments into this class
++       // Zero means "infinite" capacity i.e. the usage is not restricted
++       // +optional
++       Capacity int64
++}
+```
+
+### Kubelet
+
+Kubelet gets QoS-class resource assignment from the [PodSpec](#podspec) and
+translates these into corresponding `QoSResources` data in the CRI API. This is
+ContainerConfig message at container creation time (CreateContainerRequest) and
+PodSandboxConfig at sandbox creation time (RunPodSandboxRequest). In practice,
+there is no translation, just copying key-value pairse.
+
+Kubelet will receive the information about available QoS-class resources (the
+types of reqources and their classes) from the runtime over the CRI API (new
+Resources field in [RuntimeStatus](#runtimestatus) message). The kubelet
+updates the new QoSResources field in [NodeStatus](#nodestatus) accordingly,
+making QoS-class resources on the node visible to users and the kube-scheduler.
+
+An admission handler is added into kubelet to validate the QoS-class resource
+request against the resource availability on the node. Pod is rejected if
+sufficient resources do not exist.
+
+Invalid class assignments will cause an error in the container runtime which
+causes the corresponding CRI RuntimeService request (e.g.  RunPodSandbox or
+CreateContainer) to fail with an error. This can happen, despite the admission
+check, e.g. when the requested QoS-class resource is not available on the
+runtime (e.g. kubelet sends the pod or container start request before getting
+an update of changed QoS-class resource availability from the runtime).
+
+A feature gate QoSResources enables kubelet to interpretthe specific pod
+annotations. If the feature gate is disabled the annotations are simply ignored
+by kubelet.
+
+### API server
+
+Input validation of QoS-class resource names and class names, very similar to
+labels is implemented: keys (`QoSResourceName`) and values must be non-empty,
+less than 64 characters long, must start and end with an alphanumeric character
+and may contain only alphanumeric characters, dashes, underscores or dots (`-`,
+`_` or `.`). Also similar to labels, a namespace prefix (FQDN subdomain separated
+with a slash) in the key is allowed (e.g. `vendor/qos-resource`).
```

Comment:
```suggestion
with a slash) in the key is allowed (e.g. `vendor.example/qos-resource`).
```

We expect subdomains in this kind of key to contain at least one `.`

Reply:
Fixed